### PR TITLE
[NETBEANS-2349] Convert switch typecast case to switch expression

### PR DESCRIPTION
**Pull request by vikasprabhakar**
Convert to switch expression for following scenario (return case):

Actual Code:

void test(Object o1, Object o2, Object o3, int n) {
String result;
switch(n) { 
case 1: 
result = (String) o1;
break; 
case 2: 
result = (String) o2;
break; 
default: 
result = (String) o3;
break; 
}
System.out.println(result);
} 

After Fix with hints Hint : 'Convert to switch expression'

void test(Object o1, Object o2, Object o3, int n) {
String result;
result = (String)( switch (n){ 
case 1 -> o1; 
case 2 -> o2; 
default -> o3; }
);
System.out.println(result);
}

### CODE REVIEW

File: java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java [hunk 1]
```diff
@@ -3164,6 +3165,7 @@ public static void performRewriteRuleSwitch(JavaFix.TransformationContext ctx, T
         boolean switchExpressionFlag = st.getKind().toString().equals(TreeShims.SWITCH_EXPRESSION);
         if (switchExpressionFlag) {
```
Comment (rtaneja1) [inline java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java]: Rename this to indicate flag for rule/simplified?

File: java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java [hunk 2]
```diff
@@ -3265,16 +3277,34 @@ public Void visitIdentifier(IdentifierTree node, Void p) {
...
+            case SIMPLIFIED_WITH_RETURN_CASE:
+                et = (ExpressionTree) make.SwitchExpression(TreeShims.getExpressions(st).get(0), newCases);
+                if(typeCast != null){
+                    et = make.Parenthesized(et);
+                    et = make.TypeCast(typeCast, et);
+                }
+                wc.rewrite(st, make.Return(et));
+                break;
+            case SIMPLIFIED_WITH_CAST_CASE:
+                et = (ExpressionTree) make.SwitchExpression(TreeShims.getExpressions(st).get(0), newCases);
```
Comment (rtaneja1) [inline java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java]: typeCase null check required here as well?

File: java/java.hints/test/unit/src/org/netbeans/modules/java/hints/jdk/ConvertSwitchToRuleSwitchTest.java [hunk 3]
```diff
@@ -594,6 +594,64 @@ public void testSwitch2SwitchExpressionReturnValue() throws Exception {
                               "}\n");
     }
 
+    public void testSwitch2SwitchExpressionTypeCast() throws Exception {
```
Comment (rtaneja1) [inline java/java.hints/test/unit/src/org/netbeans/modules/java/hints/jdk/ConvertSwitchToRuleSwitchTest.java]: please test nested switch expr inside return as well

File: java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java [hunk 4]
```diff
@@ -3265,16 +3274,36 @@ public Void visitIdentifier(IdentifierTree node, Void p) {
...
+                    et = make.Parenthesized(et);
+                    et = make.TypeCast(typeCastTree, et);
+                }
+                wc.rewrite(st, make.ExpressionStatement((ExpressionTree) make.Assignment((ExpressionTree) leftVariable, et)));
+                break;
+            case SWITCH_EXPRESSION_ASSIGN_CASE:
+                et = (ExpressionTree) make.SwitchExpression(TreeShims.getExpressions(st).get(0), newCases);
+                wc.rewrite(st, make.ExpressionStatement((ExpressionTree) make.Assignment((ExpressionTree) leftVariable, et)));
+                break;
+            case SWITCH_EXPRESSION:
```
Comment (sdedic) [inline java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java]: IMHO cases for `SWITCH_EXPRESSION_RETURN_CASE`, `SWITCH_EXPRESSION_CAST_CASE`, `SWITCH_EXPRESSION_ASSIGN_CASE` can be mostly merged, except the last rewrite step, which is unique for each of them.
- all are switch expressions;
- if `typeCastTree` is available, wrap in type cast
- (the case-specific statement replacement)

**Comment by sdedic** [comment]
Also filed Issues [NETBEANS-2583](https://issues.apache.org/jira/browse/NETBEANS-2583) and [NETBEANS-2584](https://issues.apache.org/jira/browse/NETBEANS-2584)

**Comment by vikasprabhakar** [comment]
> [NETBEANS-2583](https://issues.apache.org/jira/browse/NETBEANS-2583) and [NETBEANS-2584](https://issues.apache.org/jira/browse/NETBEANS-2584)

We will provide fix for these issues in separate PR.

File: java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java [hunk 5]
```diff
@@ -3244,15 +3245,23 @@ public Void visitIdentifier(IdentifierTree node, Void p) {
...
+                    switchType = SWITCH_TYPE.SWITCH_EXPRESSION_ASSIGN_CASE;
                     JCTree.JCExpressionStatement jceTree = (JCTree.JCExpressionStatement) statements.get(0);
                     body = ((JCTree.JCAssign) jceTree.expr).rhs;
-                    variable = ((JCTree.JCAssign) jceTree.expr).lhs;
+                    if (body instanceof JCTree.JCTypeCast) {
+                        switchType = SWITCH_TYPE.SWITCH_EXPRESSION_CAST_CASE;
+                    }
+                    leftVariable = ((JCTree.JCAssign) jceTree.expr).lhs;
                 }
+                if (body instanceof JCTree.JCTypeCast) {
```
[the remark this answers is not in the document]
Comment (sdedic) [reply java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java]: sorry to nicpick, but what about code like
`return ((returnType) whateverExpression)` ?
After `Tree.Kind.Return` branch, the `body` will contain `ParenthesizedTree`.

[the remark this answers is not in the document]
Comment (vikasprabhakar) [reply java/java.hints/src/org/netbeans/modules/java/hints/errors/Utilities.java]: Could you please provide the use case code for your above feedback?
I checked with following nested scenario and suggestion hints working properly.
public String test(int p) {
        String str;
        int i = 10;
        int j = 10;
        switch (i) {
            case 1:
                return (switch (j) {
                    case 1->
                        "One";
                    default->
                        "Default";
                });

            default:
                return "default outer";

        }
    }

**Comment by geertjanw** [comment]
Ready to be merged or not?

**Comment by arusinha** [comment]
hi @geertjanw  , I will verify the PR and will approve the PR if it is fine

**Pull request merged**